### DISCUSSION TITLE
Trouble adding ignite.distributed to my baseline training process

### DISCUSSION THREAD

**Discussion by aksg87** [pytorch/ignite#2032] [Q&A] [answered]
Hello, 

I wanted to add optimal training across 8 GPUs on a single machine. 

I followed the documentation details [here](https://pytorch.org/ignite/distributed.html) and looked at the CIFAR [example](https://github.com/pytorch/ignite/tree/master/examples/contrib/cifar10).

Is there a core list of steps required to appropriately convert Ignite to support `ignite.distributed`? The CIFAR example packs a lot into a single example.

What I tried:

- Wrapped dataloaders, model, and optimizer in the `idist.auto_*` corresponding function. 
- Added a condition of `if rank == 0` for TensorboardLogger context manager. 
- Calling training loop as:
```
if __name__ == "__main__":

    with idist.Parallel(backend=backend, nproc_per_node=8) as parallel:
        parallel.run(training, config)
```

Doing all of this, I get the following TypeError. Not sure what's causing this so any ideas are appreciated.

```
-- Process 0 terminated with the following error:
Traceback (most recent call last):
  File "/home/ec2-user/programming/pipline/segmentation-v2/env/lib/python3.8/site-packages/torch/multiprocessing/spawn.py", line 59, in _wrap
    fn(i, *args)
  File "/home/ec2-user/programming/pipline/segmentation-v2/env/lib/python3.8/site-packages/ignite/distributed/comp_models/native.py", line 272, in _dist_worker_task_fn
    fn(local_rank, *args, **kw_dict)
  File "/home/ec2-user/programming/pipline/segmentation-v2/segmentation/executors/train.py", line 298, in training
    trainer.run(train_loader, max_epochs=epochs)
  File "/home/ec2-user/programming/pipline/segmentation-v2/env/lib/python3.8/site-packages/ignite/engine/engine.py", line 702, in run
    return self._internal_run()
  File "/home/ec2-user/programming/pipline/segmentation-v2/env/lib/python3.8/site-packages/ignite/engine/engine.py", line 775, in _internal_run
    self._handle_exception(e)
  File "/home/ec2-user/programming/pipline/segmentation-v2/env/lib/python3.8/site-packages/ignite/engine/engine.py", line 469, in _handle_exception
    raise e
  File "/home/ec2-user/programming/pipline/segmentation-v2/env/lib/python3.8/site-packages/ignite/engine/engine.py", line 743, in _internal_run
    self._setup_engine()
  File "/home/ec2-user/programming/pipline/segmentation-v2/env/lib/python3.8/site-packages/ignite/engine/engine.py", line 725, in _setup_engine
    self._dataloader_iter = iter(self.state.dataloader)
  File "/home/ec2-user/programming/pipline/segmentation-v2/env/lib/python3.8/site-packages/torch/utils/data/dataloader.py", line 355, in __iter__
    return self._get_iterator()
  File "/home/ec2-user/programming/pipline/segmentation-v2/env/lib/python3.8/site-packages/torch/utils/data/dataloader.py", line 301, in _get_iterator
    return _MultiProcessingDataLoaderIter(self)
  File "/home/ec2-user/programming/pipline/segmentation-v2/env/lib/python3.8/site-packages/torch/utils/data/dataloader.py", line 914, in __init__
    w.start()
  File "/opt/miniconda3/lib/python3.8/multiprocessing/process.py", line 121, in start
    self._popen = self._Popen(self)
  File "/opt/miniconda3/lib/python3.8/multiprocessing/context.py", line 224, in _Popen
    return _default_context.get_context().Process._Popen(process_obj)
  File "/opt/miniconda3/lib/python3.8/multiprocessing/context.py", line 284, in _Popen
    return Popen(process_obj)
  File "/opt/miniconda3/lib/python3.8/multiprocessing/popen_spawn_posix.py", line 32, in __init__
    super().__init__(process_obj)
  File "/opt/miniconda3/lib/python3.8/multiprocessing/popen_fork.py", line 19, in __init__
    self._launch(process_obj)
  File "/opt/miniconda3/lib/python3.8/multiprocessing/popen_spawn_posix.py", line 47, in _launch
    reduction.dump(process_obj, fp)
  File "/opt/miniconda3/lib/python3.8/multiprocessing/reduction.py", line 60, in dump
    ForkingPickler(file, protocol).dump(obj)
TypeError: cannot pickle 'SwigPyObject' object

```

**Comment by jeffydc** [comment, marked as answer]
Hi @aksg87 , looks like you are using `spawn` method to run distributed training and the error is coming from PyTorch DataLoader unable to pickle the `SwigPyObject`. Make sure DataLoader can pickle `SwigPyObject`.

Also please try to use `launch` method to run distributed training, it is faster than `spawn` method.
```sh
# spawn method
python train.py --args (training args)
# launch method
python -m torch.distributed.launch --nproc_per_node 8 --use_env -m train.py --backend nccl --args (training args)
```
If you use `launch`, calling the training loop will become:
```py
if __name__ == "__main__":

    with idist.Parallel(backend=backend) as parallel:  # no need for `nproc_per_node` as it is handled by `torch.distributed.launch`
        parallel.run(training, config)
```

[More on `torch.distributed.launch`](https://pytorch.org/docs/stable/distributed.html#launch-utility)

**Reply by vfdev-5** [reply]
In addition to this answer, it is possible to change spawning method between "fork" and "spawn":
If "forking" does not work for a certain object, maybe you can try "spawn" : 

```python
    start_method = "spawn" 
    # or 
    # start_method = "fork" 
    with idist.Parallel(backend=backend, nproc_per_node=8, start_method=start_method) as parallel:
        parallel.run(training, config)
```

But please keep in mind that while spawning there can be a slowdown with DataLoader on the first epoch (that's why @ydcjeff was saying: "Also please try to use launch method to run distributed training, it is faster than spawn method.")

See here for details: https://github.com/pytorch/ignite/issues/1938

**Comment by vfdev-5** [comment]
@aksg87 does it work now your baseline with `idist` and `python -m torch.distributed.launch --nproc_per_node 8 --use_env` as suggested @ydcjeff ?

**Reply by aksg87** [reply]
Going to try in just a few minutes @vfdev-5  -- thanks for following up!

**Reply by aksg87** [reply]
@vfdev-5  @ydcjeff 

Everything is running with your suggestions -- thank you!  I added a template of my code below in case anyone has further suggestions. I am running everything with `python -m torch.distributed.launch --nproc_per_node 8 --use_env -m train --backend nccl` on a single machine with 8 GPUs.

Two other questions I have:
1. What is the best way to use: `idist.get_rank()` and `idist.barrier()`? Is there a general pattern that works well for most use cases? I found the cifar [example](https://github.com/pytorch/ignite/blob/master/examples/contrib/cifar10/main.py) a bit hard to follow.
2. Does `ignite.distributed` optimize data usage to the GPU? I noticed making batch sizes 1 and 4 both trained pretty fast.

I'm looking forward to trying this on multiple nodes in the future!



```
# train.py

# ... imports ...


def get_data_loaders(train_batch_size, val_batch_size):

    # ... get datasets ...

    train_loader = idist.auto_dataloader(
        train_dataset, batch_size=train_batch_size, num_workers=4, shuffle=True
    )

    val_loader = idist.auto_dataloader(
        val_dataset, batch_size=val_batch_size, num_workers=4, shuffle=False
    )

    return train_loader, val_loader


def training(local_rank, config):

    train_batch_size = config["train_batch_size"]
    val_batch_size = config["val_batch_size"]

    # ... extract config items ...

    train_loader, val_loader = get_data_loaders(
        train_batch_size, val_batch_size
    )

    model = Net(1, num_classes, filters)
    model = idist.auto_model(model)

    device = idist.device()

    optimizer = torch.optim.Adam(model.parameters(), _opt.TRAIN.LR)
    optimizer = idist.auto_optim(optimizer)

    criterion = DiceInvariant3D()

    def train_step(engine, batch):
        model.train()
        optimizer.zero_grad()
        x, y = batch["image"].to(device), batch["mask"].type(
            torch.FloatTensor
        ).to(device)

        y_pred = model(x)

        loss = criterion(y_pred, y)
        loss.backward()
        optimizer.step()

        return loss.item()


    def validation_step(engine, batch):
        model.eval()
        with torch.no_grad():
            x, y = batch["image"].to(device), batch["mask"].type(
                torch.FloatTensor
            ).to(device)

            y_pred = model(x)

            return y_pred, y

    # trainer engine with loss
    trainer = Engine(train_step)

    # create two metrics engines for train validation datasets
    evaluator = Engine(validation_step)
    train_evaluator = Engine(validation_step)

    val_metrics = {"loss": Loss(criterion)}

    for name, metric in val_metrics.items():
        metric.attach(evaluator, name)
        metric.attach(train_evaluator, name)

    trainer.logger = setup_logger("trainer")
    evaluator.logger = setup_logger("evaluator")
    train_evaluator.logger = setup_logger("train_evaluator")

    pbar = tqdm(
        initial=0,
        leave=False,
        total=len(train_loader),
        desc=f"ITERATION - loss: {0:.2f}",
    )

    @trainer.on(Events.ITERATION_COMPLETED(every=log_interval))
    def log_training_loss(engine):
        pbar.desc = f"ITERATION - loss: {engine.state.output:.2f}"
        pbar.update(log_interval)

    @trainer.on(Events.EPOCH_COMPLETED(every=val_interval) | Events.COMPLETED)
    def log_training_results(engine):
        pbar.refresh()
        train_evaluator.run(train_loader)  # expensive operation
        metrics = train_evaluator.state.metrics
        # avg_accuracy = metrics["accuracy"]
        avg_loss = metrics["loss"]
        tqdm.write(
            f"Training Results - Epoch: {engine.state.epoch}"
            f" Avg loss: {avg_loss:.2f}"
        )

    @trainer.on(Events.EPOCH_COMPLETED(every=val_interval) | Events.COMPLETED)
    def log_validation_results(engine):
        evaluator.run(val_loader)  # expensive operation
        metrics = evaluator.state.metrics
        # avg_accuracy = metrics["accuracy"]
        avg_loss = metrics["loss"]
        tqdm.write(
            f"Validation Results - Epoch: {engine.state.epoch}"
            f" Avg loss: {avg_loss:.2f}"
        )

        pbar.n = pbar.last_print_n = 0

    @trainer.on(Events.EPOCH_COMPLETED | Events.COMPLETED)
    def log_time(engine):
        tqdm.write(
            f"{trainer.last_event_name.name} took"
            f" { trainer.state.times[trainer.last_event_name.name]} seconds"
        )

    # Define a Tensorboard logger
    with TensorboardLogger(log_dir="quick-start-segmentation") as tb_logger:

        # Attach handler to plot trainer's loss at frequency per log_interval
        tb_logger.attach_output_handler(
            trainer,
            event_name=Events.ITERATION_COMPLETED(every=log_interval),
            tag="training",
            output_transform=lambda loss: {"batchloss": loss},
        )

        # Attach handler to dump evaluator's metrics every epoch completed
        for tag, evaluator in [
            ("training", train_evaluator),
            ("validation", evaluator),
        ]:  # https://labs.quansight.org/blog/2020/09/pytorch-ignite/
            tb_logger.attach_output_handler(
                evaluator,
                event_name=Events.EPOCH_COMPLETED,
                tag=tag,
                metric_names="all",
                global_step_transform=global_step_from_engine(trainer),
            )


    trainer.run(train_loader, max_epochs=epochs)
    pbar.close()


config = {
    "train_batch_size": _opt.TRAIN.BATCH_SIZE,
    "val_batch_size": _opt.VAL.BATCH_SIZE,

    # ... more config items...
}


backend = "nccl"
start_method = "spawn"


if __name__ == "__main__":

    with idist.Parallel(backend=backend) as parallel:
        parallel.run(training, config)

**Reply by vfdev-5** [reply]
@aksg87 thanks for the feedback !

> What is the best way to use: idist.get_rank() and idist.barrier()? Is there a general pattern that works well for most use cases? I found the cifar example a bit hard to follow.

In distributed computations we have `world_size` on processes doing computations. We use `get_rank` to get the index of the process to execute specific things that should be only done on one process. For example, writing files on the disk (training checkpoints, best models), printing, logging, using experiment tracking systems etc.
We use `barrier` in cases when we need to synchronize all the processes. In cifar10 example we used to download the dataset only using one process and others were waiting before creating torchvision CIFAR10 class and opening the data.

> Does ignite.distributed optimize data usage to the GPU? I noticed making batch sizes 1 and 4 both trained pretty fast.

The method `ignite.distributed.auto_dataloader` does not optimize data usage, but only scales the `batch_size` and `num_workers` arguments by `world_size`. It also applies DistributedSampler to partition the data between the processes.
To decrease GPU memory usage and run faster on GPUs with "tensor cores" (e.g. Volta, Turing and later), we can think to use automatic-mixed-precision (amp): https://pytorch.org/docs/stable/notes/amp_examples.html#automatic-mixed-precision-examples


HTH

**Reply by aksg87** [reply]
Thanks for the clarifications! @vfdev-5 

So in my template, I am assuming we should force `rank == 0` for the `TensorboardLogger` context manager block, and `setup_logger`, and `pbar` setup/close. 

I was curious about if we need to control the main engine `trainer.run(train_loader, max_epochs=epochs)` under `rank == 0` or if we have a separate main engine running for each process. Just trying to better understand the distributed workflow! thanks again

**Reply by vfdev-5** [reply]
Sure, feel free to ask questions here or on [our Discord](https://discord.gg/djZtm3EmKj).

> So in my template, I am assuming we should force rank == 0 for the TensorboardLogger context manager block, and setup_logger, and pbar setup/close.

yes, TensorBoardLogger context manager and all associated usage, logger and pbar should be in one process (say, rank=0). Same as here:
https://github.com/pytorch/ignite/blob/f5b0c5cb17cdc2a57a7a6790038e50ee652b94b2/examples/contrib/cifar10/main.py#L98-L104

Otherwise, each process will display messages or log into the filesystem which is redundant.

> I was curious about if we need to control the main engine trainer.run(train_loader, max_epochs=epochs) under rank == 0 or if we have a separate main engine running for each process.

Currently, here we are doing data distributed parallelism, it means each process has a copy of model, optimizer, loss function and process its own chunk of data. Every process should train the model and pytorch DDP wrapper handles gradients sync and all necessary things that we need to train one model instead of N unrelated models.

In this case `trainer.run` should be called by all processes. Same as here: 
https://github.com/pytorch/ignite/blob/f5b0c5cb17cdc2a57a7a6790038e50ee652b94b2/examples/contrib/cifar10/main.py#L129

**Reply by aksg87** [reply]
Thanks @vfdev-5 -- I'll check out the discord for discussions!

Btw -- it would be really cool if the context managers, loggers, etc. could be wrapped in an ignite.distributed.auto_* object to handle things behind the scenes.

**Reply by vfdev-5** [reply]
> Btw -- it would be really cool if the context managers, loggers, etc. could be wrapped in an ignite.distributed.auto_* object to handle things behind the scenes.

@aksg87 thanks for the feedback, I agree that some simplifications on logging, checkpointing could be done behind the scene. We are currently working on a higher level API that expose the engine for flexibility and should provide out-of-the-box DDP support, checkpointing, evaluation and logging...
cc @KickItLikeShika
